packages/open-next/src/adapters/routing.ts

```diff
@@ -0,0 +1,104 @@
+import { InternalEvent } from './event-mapper.js';
+import { debug } from './logger.js';
+import {
+  RedirectDefinition,
+  RewriteDefinition,
+  RewriteMatcher,
+  RoutesManifest,
+} from './next-types.js';
+
+const redirectMatcher =
+  (
+    headers: Record<string, string>,
+    cookies: Record<string, string>,
+    query: Record<string, string | string[]>
+  ) =>
+  (redirect: RewriteMatcher) => {
+    switch (redirect.type) {
+      case 'header':
+        console.log(
+          'redirect-header',
+          headers?.[redirect.key.toLowerCase()],
+          redirect.value
+        );
+        return (
+          headers?.[redirect.key.toLowerCase()] &&
+          new RegExp(redirect.value ?? '').test(
+            headers[redirect.key.toLowerCase()] ?? ''
+          )
+        );
+      case 'cookie':
+        return (
+          cookies?.[redirect.key] &&
+          new RegExp(redirect.value ?? '').test(cookies[redirect.key] ?? '')
+        );
+      case 'query':
+        return query[redirect.key] && Array.isArray(redirect.value)
+          ? redirect.value.reduce(
+              (prev, current) => prev || new RegExp(current),
+              true
+            )
+          : new RegExp(redirect.value ?? '').test(
+              (query[redirect.key] as string | undefined) ?? ''
+            );
+      case 'host':
+        return (
+          headers?.host && new RegExp(redirect.value ?? '').test(headers.host)
+        );
+      default:
+        return false;
+    }
+  };
+
+export function handleRewrites<T extends RewriteDefinition>(
+  internalEvent: InternalEvent,
+  rewrites: T[]
+) {
+  const { rawPath, headers, query, cookies } = internalEvent;
+  const matcher = redirectMatcher(headers, cookies, query);
+  const rewrite = rewrites.find(
+    (route) =>
+      new RegExp(route.regex).test(rawPath) &&
+      (route.has
+        ? route.has.reduce((acc, cur) => {
+            if (acc === false) return false;
+            return matcher(cur);
+          }, true)
+        : true) &&
+      (route.missing
+        ? route.missing.reduce((acc, cur) => {
+            if (acc === false) return false;
+            return !matcher(cur);
+          }, true)
+        : true)
+  );
+
+  const urlQueryString = new URLSearchParams(query).toString();
+  if (rewrite) {
+    debug('rewrite', { rewrite, urlQueryString });
+  }
+  // Do we need to replace anything else here?
+  const rewrittenUrl =
+    rewrite?.destination.replace(/:[^\/]+/g, '[$1]') ?? rawPath;
```

Comment:
```suggestion
    rewrite?.destination.replace(/:([^\/]+)/g, '[$1]') ?? rawPath;
```
You forgot the grouping `()`, otherwise, it'll replace the word with a literal `$1`

Reply:
Yeah i noticed that, testing this right now

Reply:
This wasn't working finally, we need to replace the `rawPath` here, i used https://github.com/pillarjs/path-to-regexp. This is the lib they use internally inside next

Reply:
Hmm I'm getting 502's on prefetch now

Reply:
It's likely that it comes from here https://github.com/conico974/open-next/blob/b3060a37960e4a1305fe82d125d5003c7e0897be/packages/open-next/src/adapters/routing.ts#L74-L85. Could you try adding some log here to see what could be the cause

Reply:
Yup I'll do that as soon as I get back from shopping.

Reply:
I added a bunch of logs after each line and it appears to crash here: https://github.com/conico974/open-next/blob/b3060a37960e4a1305fe82d125d5003c7e0897be/packages/open-next/src/adapters/routing.ts#L82C8-L82C8

Reply:
We need the interception path here otherwise it will just redirect, but looking at the error, it might be this https://github.com/vercel/next.js/blob/da043b5535bf50f699c928d26052aba07cb243bc/packages/next/src/shared/lib/router/utils/path-match.ts#L65. Could you look what your params look like, and if they have number keys, maybe try to remove them

Reply:
I updated my comment w/ additional info.

Reply:
I should have read the docs a little more... https://github.com/pillarjs/path-to-regexp#unnamed-parameters. So in `compile` it treats `(.)` as an unnamed parameter, and so it expects it to be present when we use it in `toDestination`. see https://github.com/pillarjs/path-to-regexp/blob/1cbb9f3d9c3bff97298ec45b1bb2b0beb879babf/src/index.spec.ts#L2892
We actually need the interception path (i.e. `(.)|(..)|(...)`) to be returned from our function so that it return the correct route. I think we need to escape those interception path for when we use `toDestination`  and then replace those escaped path with  `(.)|(..)|(...)` so that we return the expected path

Reply:
Hmm but isn't that (.) regex, and not directory paths?

Reply:
I added a util function in utils.ts:
```
const replacePathRegex = /(\(?\w+\)?\/)?\(\.{2}\)\/?/;
export function resolvePaths(path: string) {
  const rootIndex = path.indexOf("(...)");
  // If path contains (...), replace everything before it, including itself
  if (~rootIndex) {
    path = path.substring(rootIndex + 5);
  }
  // Remove (.), since that resolves to nothing
  path = path.replace(/\(\.\)/g, "");

  // Resolves each (..) path by consuming the path before it
  // eg: /page/(..)[dynamic] => /[dynamic]
  while (path.match(/\(\.{2}\)/)) {
    path = path.replace(replacePathRegex, "");
  }
  return path;
}
```

The in routing.ts, I resolved the source & destination:
```
    const toDestination = compile(resolvePaths(rewrite?.destination) ?? "");
    const fromSource = match(resolvePaths(rewrite?.source) ?? "");
    ```

Reply:
Nvm, the modal's param.title is messed up for some reason, it has `%3Atitle` from (params.title)